File: detox/ios/DetoxTester/DetoxTester/Actions/XCUIElement/ElementAttributes.swift
//
//  ElementAttributes.swift (DetoxTesterApp)
//  Created by Asaf Korem (Wix.com) on 2022.
//

import Foundation
import DetoxInvokeHandler

/// Represents various attributes of an element.
struct ElementAttributes: Codable, Equatable {
  /// The text value of any textual element.
  let text: String?

  /// A succinct label in a localized string that identifies the accessibility element.
  let label: String?

  /// The placeholder text value of an input element.
  let placeholder: String?

  /// Whether the element is enabled for user interaction.
  let enabled: Bool?

  /// The identifier of the element.
  let identifier: String?

  /// Indicates whether the element is visible.
  let visible: Bool?

  /// The value of the element, where applicable.
  ///
  /// - Note: For example, the position of a slider, or whether a checkbox has been marked.
  let value: String?

  /// The activation point of the element, in element coordinate space.
  let activationPoint: Point?

  /// The activation point of the element (`activationPoint`), normalized ([0.0, 1.0]) in the
  /// element's bounds.
  let normalizedActivationPoint: Point?

  /// Indicates whether the system can compute a hit point for the element.
  let hittable: Bool?

  /// The frame of the element, in screen coordinate space.
  let frame: Rect?

  /// The frame of the element, in container coordinate space. Matches
  /// `accessibilityFrameInContainerSpace`.
  let elementFrame: Rect?

  /// The bounds of the element, in element coordinate space.
  let elementBounds: Rect?

  /// The safe area insets of the element, in element coordinate space.
  let safeAreaInsets: EdgeInsets?

  /// The safe area bounds of the element, in element coordinate space.
  let elementSafeBounds: Rect?

  /// The date of the element (in case the element is a date picker).
  let date: Date?

  /// The normalized slider position (in case the element is a slider).
  let normalizedSliderPosition: CGFloat?

  /// The content offset (in case the element is a scroll view).
  let contentOffset: CGFloat?

  /// The content inset (in case the element is a scroll view).
  let contentInset: CGFloat?

  /// The adjusted content inset (in case the element is a scroll view).
  let adjustedContentInset: CGFloat?

  /// The type of the element.
  let type: String

  /// A localized string that contains a brief description of the result of performing an action
  /// on the accessibility element.
  let hint: String?

  /// Indicates whether a selectable element is currently selected or not.
  let selected: Bool?

  /// Indicates whether the element is focused.
  let focused: Bool?
}

extension ElementAttributes {
  init(text: String?, accessibilityLabel: String?,
       placeholderValue: String?, isEnabled: Bool?, accessibilityIdentifier: String?,
       isVisible: Bool?, accessibilityValue: String?, activationPoint: CGPoint?,
       normalizedActivationPoint: CGPoint?, isHittable: Bool?, frame: CGRect?,
       elementFrame: CGRect?, elementBounds: CGRect?, safeAreaInsets: UIEdgeInsets?,
       elementSafeBounds: CGRect?, datePickerDate: Date?, normalizedSliderPosition: CGFloat?,
       scrollViewContentOffset: CGFloat?, scrollViewContentInset: CGFloat?,
       scrollViewAdjustedContentInset: CGFloat?, elementType: XCUIElement.ElementType,
       accessibilityHint: String?, isSelected: Bool?, isFocused: Bool?) {
    self.text = text
    self.label = accessibilityLabel
    self.placeholder = placeholderValue
    self.enabled = isEnabled
    self.identifier = accessibilityIdentifier
    self.visible = isVisible
    self.value = accessibilityValue
    self.activationPoint = activationPoint?.codable
    self.normalizedActivationPoint = normalizedActivationPoint?.codable
    self.hittable = isHittable
    self.frame = frame?.codable
    self.elementFrame = elementFrame?.codable
    self.safeAreaInsets = safeAreaInsets?.codable
    self.elementBounds = elementBounds?.codable
    self.elementSafeBounds = elementSafeBounds?.codable
    self.date = datePickerDate
    self.normalizedSliderPosition = normalizedSliderPosition
    self.contentOffset = scrollViewContentOffset
    self.contentInset = scrollViewContentInset
    self.adjustedContentInset = scrollViewAdjustedContentInset
    self.type = elementType.asString()
    self.hint = accessibilityHint
    self.selected = isSelected
    self.focused = isFocused
  }
}

/// Represents a point in 2D space.
struct Point: Codable, Equatable {
  /// The x coordinate of the point.
  let x: CGFloat

  /// The y coordinate of the point.
  let y: CGFloat
}

/// Represents a rectangle in 2D space.
struct Size: Codable, Equatable {
  /// The width of the rectangle.
  let height: CGFloat

  /// The height of the rectangle.
  let width: CGFloat
}

/// Represents a rectangle in 2D space.
struct Rect: Codable, Equatable {
  /// The origin of the rectangle.
  let origin: Point

  /// The size of the rectangle.
  let size: Size
}

/// Represents the insets of a rectangle.
struct EdgeInsets: Codable, Equatable {
  /// The top inset of the rectangle.
  let top: CGFloat

  /// The left inset of the rectangle.
  let bottom: CGFloat

  /// The right inset of the rectangle.
  let left: CGFloat

  /// The bottom inset of the rectangle.
  let right: CGFloat
}

extension CGPoint {
  /// A `Point` representation of the `CGPoint`.
  var codable: Point {
    .init(x: x, y: y)
  }
}

extension CGSize {
  /// A `Size` representation of the `CGSize`.
  var codable: Size {
    .init(height: height, width: width)
  }
}

extension CGRect {
  /// A `Rect` representation of the `CGRect`.
  var codable: Rect {
    .init(
      origin: .init(x: origin.x, y: origin.y),
      size: .init(height: size.height, width: size.width)
    )
  }
}

extension UIEdgeInsets {
  /// An `EdgeInsets` representation of the `UIEdgeInsets`.
  var codable: EdgeInsets {
    .init(top: top, bottom: bottom, left: left, right: right)
  }
}

File: packages/client/src/types/ElementAttributes.ts
import { Point2D } from '../common';

/** Shared element attributes between iOS and Android. */
export interface ElementAttributes {
  /**
   * Whether or not the element is enabled for user interaction.
   */
  enabled: boolean;
  /**
   * The identifier of the element. Matches accessibilityIdentifier on iOS, and the main view tag, on Android - both commonly holding the component's test ID in React Native apps.
   */
  identifier: string;
  /**
   * Whether the element is visible. On iOS, visibility is calculated for the activation point. On Android, the attribute directly holds the value returned by View.getLocalVisibleRect()).
   */
  visible: boolean;
  /**
   * The text value of any textual element.
   */
  text?: string;
  /**
   * The label of the element. Largely matches accessibilityLabel for ios, and contentDescription for android.
   * Refer to Detox's documentation (`toHaveLabel()` subsection) in order to learn about caveats associated with
   * this property in React Native apps.
   */
  label?: string;
  /**
   * The placeholder text value of the element. Matches hint on android.
   */
  placeholder?: string;
  /**
   * The value of the element, where applicable.
   * Matches accessibilityValue, on iOS.
   * For example: the position of a slider, or whether a checkbox has been marked (Android).
   */
  value?: unknown;
}

export interface IosElementAttributes extends ElementAttributes {
  /**
   * The [activation point]{@link https://developer.apple.com/documentation/objectivec/nsobject/1615179-accessibilityactivationpoint} of the element, in element coordinate space.
   */
  activationPoint: Point2D;
  /**
   * The activation point of the element, in normalized percentage ([0.0, 1.0]).
   */
  normalizedActivationPoint: Point2D;
  /**
   * Whether the element is hittable at the activation point.
   */
  hittable: boolean;
  /**
   * The frame of the element, in screen coordinate space.
   */
  frame: IosElementAttributeFrame;
  /**
   * The frame of the element, in container coordinate space.
   */
  elementFrame: IosElementAttributeFrame;
  /**
   * The bounds of the element, in element coordinate space.
   */
  elementBounds: IosElementAttributeFrame;
  /**
   * The safe area insets of the element, in element coordinate space.
   */
  safeAreaInsets: IosElementAttributeInsets;
  /**
   * The safe area bounds of the element, in element coordinate space.
   */
  elementSafeBounds: IosElementAttributeFrame;
  /**
   * The date of the element (if it is a date picker).
   */
  date?: string;
  /**
   * The normalized slider position (if it is a slider).
   */
  normalizedSliderPosition?: number;
  /**
   * The content offset (if it is a scroll view).
   */
  contentOffset?: Point2D;
  /**
   * The content inset (if it is a scroll view).
   */
  contentInset?: IosElementAttributeInsets;
  /**
   * The adjusted content inset (if it is a scroll view).
   */
  adjustedContentInset?: IosElementAttributeInsets;
  /**
   * @example "<CALayer: 0x600003f759e0>"
   */
  layer: string;
}

interface IosElementAttributeFrame {
  y: number;
  x: number;
  width: number;
  height: number;
}

interface IosElementAttributeInsets {
  right: number;
  top: number;
  left: number;
  bottom: number;
}

export interface AndroidElementAttributes extends ElementAttributes {
  /**
   * The OS visibility type associated with the element: visible, invisible or gone.
   */
  visibility: 'visible' | 'invisible' | 'gone';
  /**
   * Width of the element, in pixels.
   */
  width: number;
  /**
   * Height of the element, in pixels.
   */
  height: number;
  /**
   * Elevation of the element.
   */
  elevation: number;
  /**
   * Alpha value for the element.
   */
  alpha: number;
  /**
   * Whether the element is the one currently in focus.
   */
  focused: boolean;
  /**
   * The text size for the text element.
   */
  textSize?: number;
  /**
   * The length of the text element (character count).
   */
  length?: number;
}
